### PR TITLE
tBTC v1 sunsetting: VendingMachineV2 and VendingMachineV3

### PR DESCRIPTION
**Pull request by pdyraga**
There are two smart contracts supporting the tBTC v1 sunsetting process: `VendingMachineV2` and `VendingMachineV3`.

`VendingMachineV2` is used to exchange tBTC v1 to tBTC v2 in a 1:1 ratio. The redeemer selected by the DAO based on the ["TIP-027b tBTC v1: The Sunsetting"](https://forum.threshold.network/t/tip-027b-tbtc-v1-the-sunsettening/357)  proposal will deposit tBTC v2 tokens into VendingMachineV2 so that outstanding tBTC v1 token owners can upgrade to tBTC v2 tokens. The redeemer will withdraw the tBTC v1 tokens deposited into the contract to perform tBTC v1 redemptions. The redeemer may decide to withdraw their deposited tBTC v2 at any moment in time. The amount withdrawable is lower than the amount deposited in case tBTC v1 was exchanged for tBTC v2.

`VendingMachineV3` is used to exchange tBTC v1 to tBTC v2 in a 1:1 ratio after the tBTC v1 bridge sunsetting is completed. Since tBTC v1 bridge will no longer work, it will not be possible to use tBTC v1 tokens to perform BTC redemptions. This contract allows tBTC v1 owners to upgrade to tBTC v2 without any deadline. This way, tBTC v1 tokens left on the market will be always backed by Bitcoin. The Treasury Guild must deposit tBTC v2 into the contract and the Threshold Council is allowed to withdraw tBTC v2 only if tBTC v2 left in this contract is enough to cover the upgrade of all tBTC v1 left on the market.

Info: Note that I used "tBTC" instead of "TBTC" when talking about the token in the documentation and in revert messages. There was a discussion in the DAO to use "tBTC" to talk both about the system and the token.

### CODE REVIEW

**Comment by pdyraga** [comment]
Please hold on with review of this PR, I need to incorporate changes based on the new approach suggested by the DAO Treasury Guild.

**Comment by pdyraga** [comment]
This is ready for review now!

File: solidity/contracts/bridge/VendingMachineV2.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+///         deposited in case tBTC v1 was exchanged for tBTC v2.
+contract VendingMachineV2 is Ownable {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for TBTC;
+
+    IERC20 public immutable tbtcV1;
+    TBTC public immutable tbtcV2;
+
+    event Exchanged(address indexed to, uint256 amount);
+    event Deposited(address from, uint256 amount);
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV2.sol:30]: We will have just one `Deposited` event in the lifetime of this contract in practice and the intention was to show it is not an often emitted event and we do not really care who deposited.

I can make it `indexed` though, it's a small gas difference.

File: solidity/contracts/bridge/VendingMachineV2.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    event Withdrawn(address token, address to, uint256 amount);
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV2.sol:31]: Same answer as [here](https://github.com/keep-network/tbtc-v2/pull/525#discussion_r1122744985). Let me know what you think and I can make the change.

File: solidity/contracts/bridge/VendingMachineV2.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    /// @param amount The amount of tBTC v2 to deposit into the contract.
+    function depositTbtcV2(uint256 amount) external {
+        emit Deposited(msg.sender, amount);
+        tbtcV2.safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    /// @notice Allows the contract owner to withdraw tokens. This function is
+    ///         used in two cases: 1) when the redeemer wants to redeem tBTC v1
+    ///         tokens to perform tBTC v2 redemptions; 2) when the deadline for
+    ///         tBTC v1 > tBTC v2 exchange passed and the redeemer wants their
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV2.sol]: b5c476d605b037dea6e21a7b9ab4bff06545e90b

File: solidity/contracts/bridge/VendingMachineV2.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    function _exchange(address tokenOwner, uint256 amount) internal {
+        emit Exchanged(tokenOwner, amount);
+        tbtcV1.safeTransferFrom(tokenOwner, address(this), amount);
+
+        require(
+            tbtcV2.balanceOf(address(this)) >= amount,
+            "Not enough tBTC v2 available in the Vending Machine"
+        );
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV2.sol:106]: b5c476d605b037dea6e21a7b9ab4bff06545e90b

File: solidity/contracts/bridge/VendingMachineV3.sol [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    function _exchange(address tokenOwner, uint256 amount) internal {
+        emit Exchanged(tokenOwner, amount);
+        tbtcV1.safeTransferFrom(tokenOwner, address(this), amount);
+
+        require(
+            tbtcV2.balanceOf(address(this)) >= amount,
+            "Not enough tBTC v2 available in the Vending Machine"
+        );
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV3.sol:126]: b5c476d605b037dea6e21a7b9ab4bff06545e90b

File: solidity/contracts/bridge/VendingMachineV2.sol [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    ///         used in two cases: 1) when the redeemer wants to redeem tBTC v1
+    ///         tokens to perform tBTC v2 redemptions; 2) when the deadline for
+    ///         tBTC v1 > tBTC v2 exchange passed and the redeemer wants their
+    ///         tBTC v2 back.
+    /// @dev This function is for the redeemer. This is NOT a function for
+    ///      tBTC v1 token holders.
+    /// @param token The address of a token to withdraw.
+    /// @param recipient The address which should receive withdrawn tokens.
+    /// @param amount The amount to withdraw.
+    function withdrawFunds(
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV2.sol:94]: The redeemer is the owner of `VendingMachineV2. I clarified it in 09515378fc223feb554e2b7dc98b98fbb9872ccb.

File: solidity/contracts/bridge/VendingMachineV3.sol [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    using SafeERC20 for IERC20;
+    using SafeERC20 for TBTC;
+
+    IERC20 public immutable tbtcV1;
+    TBTC public immutable tbtcV2;
+
+    event Exchanged(address indexed to, uint256 amount);
+    event Deposited(address from, uint256 amount);
+    event TbtcV2Withdrawn(address to, uint256 amount);
+    event FundsRecovered(address token, address to, uint256 amount);
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV3.sol:32]: Same answer as [here](https://github.com/keep-network/tbtc-v2/pull/525#discussion_r1122744985). Let me know what you think and I can make the change.

File: solidity/contracts/bridge/VendingMachineV3.sol [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    /// @param recipient The address which should receive withdrawn tokens.
+    /// @param amount The amount to withdraw.
+    function withdrawTbtcV2(address recipient, uint256 amount)
+        external
+        onlyOwner
+    {
+        require(
+            tbtcV1.totalSupply() <= tbtcV2.balanceOf(address(this)) - amount,
+            "tBTC v1 must not be left unbacked"
+        );
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV3.sol:97]: >  Maybe I'm missing something but the above means the v1 circulating supply (not locked in the vending machine yet) must be backed by the same amount of v2 tokens deposited in the vending machine.

Yes, this is correct. The Treasury Guild will mint TBTC v2 in the amount equal to the existing supply of TBTC v1 and will deposit it into this contract.

If we wanted to enforce the amount, we should probably do it in the `depositTbtcV2` function but given the minting is a multi-step process and the v1 supply can change, I decided to keep it simple.

> The ultimate goal is that the vending machine controls all v1 supply right? Do we plan to burn those tokens then?

They may be used later for taking auctions from v1 and recovering some ETH.

The current DAO plan for `VendingMachineV3` is:

1. The Treasury Guild deposits an amount of tBTC v2 matching outstanding v1 supply into the Vending Machine.
2. There are three new nodes set by the DAO that hold outstanding deposits after the churning phase. Let's call them "Trinity"
3.  The Trinity seizes BTC.
16. The Trinity loses ~102% collateral of ETH.
17. The DAO refunds The Trinity for the difference, so about 2% difference between BTC seized and the collateral lost.
18. The tBTC v1 left is backed 1:1 against tBTC v2 in the VendingMachineV3
19. Anyone who has tBTC v1 left can update to tBTC v2 indefinitely.
20. The DAO can withdraw all tBTC v1 that are deposited into the vending machine or an amount of tBTC v2 not exceeding the outstanding tBTC v1 supply (can not make tBTC v1 unbacked by withdrawing)
21. The DAO may withdraw tBTC v1 from the `VendingMachineV3` to trigger redemptions in v1. The redemptions will fail but they will be able to get some of the bonded ETH this way.

File: solidity/contracts/bridge/VendingMachineV3.sol [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+        emit TbtcV2Withdrawn(recipient, amount);
+        tbtcV2.safeTransfer(recipient, amount);
+    }
+
+    /// @notice Allows the governance to recover ERC20 sent to this contract
+    ///         by mistake or tBTC v1 locked in the contract to exchange to
+    ///         tBTC v2. No tBTC v2 can be withdrawn using this function.
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/VendingMachineV3.sol:105]: Explained [in this comment](https://github.com/keep-network/tbtc-v2/pull/525#discussion_r1122824603). Let me know if it makes sense now.

File: solidity/contracts/bridge/VendingMachineV2.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/VendingMachineV2.sol:31]: If so, I think it's ok to have them non-indexed.

File: solidity/contracts/bridge/VendingMachineV3.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/VendingMachineV3.sol:97]: Ok, all clear now! Thanks for the explanation!

File: solidity/contracts/bridge/VendingMachineV3.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/VendingMachineV3.sol:105]: Yep, makes sense!

**Review by lukasz-zimnoch — APPROVED** [review]

**Pull request merged**